rllib/examples/algorithms/classes/vpg.py

```diff
@@ -74,34 +111,66 @@ def training_step(self) -> ResultDict:
         """
         # Sample.
         with self.metrics.log_time((TIMERS, ENV_RUNNER_SAMPLING_TIMER)):
-            episodes, env_runner_results = synchronous_parallel_sample(
-                worker_set=self.env_runner_group,
-                max_env_steps=self.config.total_train_batch_size,
-                sample_timeout_s=self.config.sample_timeout_s,
-                _uses_new_env_runners=True,
-                _return_metrics=True,
-            )
+            episodes, env_runner_results = self._sample_episodes()
+        # Merge results from n parallel sample calls into self's metrics logger.
         self.metrics.merge_and_log_n_dicts(env_runner_results, key=ENV_RUNNER_RESULTS)
 
+        # Just for demonstration purposes, log the number of time steps sampled in this
+        # `training_step` round.
+        # Mean over a window of 100:
+        self.metrics.log_value(
+            "episode_timesteps_sampled_mean_win100",
+            sum(map(len, episodes)),
+            reduce="mean",
+            window=100,
+        )
+        # Exponential Moving Average (EMA) with coeff=0.1:
+        self.metrics.log_value(
+            "episode_timesteps_sampled_ema",
+            sum(map(len, episodes)),
+            ema_coeff=0.1,  # <- weight of new value; weight of old avg=1.0-ema_coeff
+        )
+
         # Update model.
         with self.metrics.log_time((TIMERS, LEARNER_UPDATE_TIMER)):
             learner_results = self.learner_group.update_from_episodes(
                 episodes=episodes,
                 timesteps={
                     NUM_ENV_STEPS_SAMPLED_LIFETIME: (
-                        self.metrics.peek(NUM_ENV_STEPS_SAMPLED_LIFETIME)
+                        self.metrics.peek(
+                            (ENV_RUNNER_RESULTS, NUM_ENV_STEPS_SAMPLED_LIFETIME)
+                        )
                     ),
                 },
             )
-        self.metrics.log_dict(learner_results, key=LEARNER_RESULTS)
+        # Merge results from m parallel update calls into self's metrics logger.
+        self.metrics.merge_and_log_n_dicts(learner_results, key=LEARNER_RESULTS)
 
         # Sync weights.
         with self.metrics.log_time((TIMERS, SYNCH_WORKER_WEIGHTS_TIMER)):
             self.env_runner_group.sync_weights(
                 from_worker_or_learner_group=self.learner_group,
-                policies=list(set(learner_results.keys()) - {ALL_MODULES}),
                 inference_only=True,
             )
 
-        # Reduce and return collected metrics.
-        return self.metrics.reduce()
+    def _sample_episodes(self):
+        # How many episodes to sample from each EnvRunner?
+        num_episodes_per_env_runner = (
+            self.config.num_episodes_per_train_batch // self.config.num_env_runners
+        )
+        # Send parallel remote requests to sample and get the metrics.
+        sampled_data = self.env_runner_group.foreach_env_runner(
```

Comment:
Nice example for sampling episodes!

Reply:
Yeah, I also like how simple it turned out. It used to be so much more complicated to write your own algo. Now, it's pretty easy to do so thanks to a couple of new stack enhancements:
* Unified MetricsLogger
* EnvRunner API is simpler than RolloutWorkers with the `num_timesteps` vs `num_episodes` explicitly sent into `sample()`.
* Episodes as data-vehicles, not batches.
* Loss function is separated from model.
* Algo defines, which RLModule APIs it needs. Other than that, users can provide any(!) RLModule they want. Here, VPG does not require any APIs :)